Backport fix for "Magento Framework Escaper - Critical log with special symbols" for Magento <= 2.2.7
Error: main.CRITICAL: DOMDocument::loadHTML(): htmlParseEntityRef: no name in Entity, line: 1

Bug confirmed present in 2.2.7(HAAR)

Original issue: https://github.com/magento/magento2/issues/13269
PR: https://github.com/magento/magento2/commit/15c7b1ba6db51af92b77896c547d2bc0a60c64a6#diff-08994a105f18db0aaf3779ce22b1805f

diff --git a/Escaper.php b/Escaper.php
index fff11dd..7a406da 100644
--- a/Escaper.php
+++ b/Escaper.php
@@ -77,6 +77,7 @@ class Escaper
                         throw new \Exception($errorString, $errorNumber);
                     }
                 );
+                $data = $this->prepareUnescapedCharacters($data);
                 $string = mb_convert_encoding($data, 'HTML-ENTITIES', 'UTF-8');
                 try {
                     $domDocument->loadHTML(
@@ -105,6 +106,19 @@ class Escaper
         return $result;
     }
 
+    /**
+     * Used to replace characters, that mb_convert_encoding will not process
+     *
+     * @param string $data
+     * @return string|null
+     */
+    private function prepareUnescapedCharacters($data)
+    {
+        $patterns = ['/\&/u'];
+        $replacements = ['&amp;'];
+        return preg_replace($patterns, $replacements, $data);
+    }
+
     /**
      * Remove not allowed tags
      *
